oauth2_http/java/com/google/auth/oauth2/AwsCredentials.java

```diff
@@ -0,0 +1,348 @@
+    // The targeted region is required to generate the signed request. The regional
+    // endpoint must also be used.
+    String region = getAwsRegion();
+
+    AwsSecurityCredentials credentials = getAwsSecurityCredentials();
+
+    // Generate the signed request to the AWS STS GetCallerIdentity API.
+    Map<String, String> headers = new HashMap<>();
+    headers.put("x-goog-cloud-target-resource", audience);
+
+    AwsRequestSigner signer =
+        AwsRequestSigner.newBuilder(
+                credentials,
+                "POST",
+                ((AwsCredentialSource) credentialSource)
+                    .regionalCredentialVerificationUrl.replace("{region}", region),
+                region)
+            .setAdditionalHeaders(headers)
+            .build();
+
+    AwsRequestSignature awsRequestSignature = signer.sign();
+    return buildSubjectToken(awsRequestSignature);
+  }
+
+  /** Clones the AwsCredentials with the specified scopes. */
+  @Override
+  public GoogleCredentials createScoped(Collection<String> newScopes) {
+    return new AwsCredentials(
+        transportFactory,
+        audience,
+        subjectTokenType,
+        tokenUrl,
+        (AwsCredentialSource) credentialSource,
+        tokenInfoUrl,
+        serviceAccountImpersonationUrl,
+        quotaProjectId,
+        clientId,
+        clientSecret,
+        newScopes);
+  }
+
+  private String retrieveResource(String url, String resourceName) throws IOException {
+    try {
+      HttpRequestFactory requestFactory = transportFactory.create().createRequestFactory();
+      HttpRequest request = requestFactory.buildGetRequest(new GenericUrl(url));
+      HttpResponse response = request.execute();
+      return response.parseAsString();
+    } catch (IOException e) {
+      throw new IOException(String.format("Failed to retrieve AWS %s.", resourceName), e);
+    }
+  }
+
+  private String buildSubjectToken(AwsRequestSignature signature)
+      throws UnsupportedEncodingException {
+    Map<String, String> canonicalHeaders = signature.getCanonicalHeaders();
+    List<GenericJson> headerList = new ArrayList<>();
+    for (String headerName : canonicalHeaders.keySet()) {
+      headerList.add(formatTokenHeaderForSts(headerName, canonicalHeaders.get(headerName)));
+    }
+
+    headerList.add(formatTokenHeaderForSts("Authorization", signature.getAuthorizationHeader()));
+
+    // The canonical resource name of the workload identity pool provider.
+    headerList.add(formatTokenHeaderForSts("x-goog-cloud-target-resource", audience));
+
+    GenericJson token = new GenericJson();
+    token.setFactory(OAuth2Utils.JSON_FACTORY);
+
+    token.put("headers", headerList);
+    token.put("method", signature.getHttpMethod());
+    token.put(
+        "url",
+        ((AwsCredentialSource) credentialSource)
+            .regionalCredentialVerificationUrl.replace("{region}", signature.getRegion()));
+    return URLEncoder.encode(token.toString(), "UTF-8");
```

Comment:
the exception really shouldn't happen here unless the VM is broken. Ine if the few cases whee I'm OK converting to a `new RuntimeException("this can't happen", e)`